ldap/servers/plugins/replication/cl5_api.c

```diff
@@ -1705,6 +1739,213 @@ cl5DBData2Entry(const char *data, PRUint32 len __attribute__((unused)), CL5Entry
+                /* Lets assume that:
+                 * it is a CL5_ error already logged by the callback
+                 */
+                errmsg = NULL;
+                clrc = rc;
+            } else {
+                clrc = CL5_DB_ERROR;
+            }
+            break;
+    }
+    if (errmsg) {
+        slapi_log_err(loglvl, repl_plugin_name_cl,
+                      "%s on changelog %s; db error - %d %s\n",
+                      errmsg, cldb->ident, rc, dblayer_strerror(rc));
+    }
+    return clrc;
+}
+
+/* Iterate over a changelog cursor */
+int
+_cl5Iterate(cldb_Handle *cldb, dbi_iterate_cb_t *action_cb, DBLCI_CTX *dblcictx, PRBool readonly)
+{
+    dbi_txn_t *txnid = NULL;
+    int rc = CL5_DB_RETRY;
+    int nbtries = 0;
+    CSN *startcsn = NULL;
+
+    dblcictx->finished = PR_FALSE;
+    dblcictx->cldb = cldb;
+    while ( !slapi_is_shutting_down() &&
+            ((rc == CL5_SUCCESS && dblcictx->finished == PR_FALSE) ||
+             (rc == CL5_DB_RETRY && nbtries < MAX_RETRIES))) {
+        nbtries++;
+        dblcictx->changed.nb = 0;
+        dblcictx->seen.nb = 0;
+        if (rc == CL5_SUCCESS) {
+            /* action_cb have decided to abort/commit the txn
+             * then go on with the iteration
+             */
+            startcsn = &dblcictx->csn;
+        }
+        if (startcsn) {
+            /* Save startcsn to be able to retry at the same place */
+            dblcictx->startcsn = *startcsn;
+            startcsn = &dblcictx->startcsn;
+        }
+        /* DB txn lock accessed pages until the end of the transaction. */
+        rc = TXN_BEGIN(cldb, NULL, &txnid, readonly);
+        if (rc != 0) {
+            rc = _cl5Dberror(cldb, rc, "_cl5Iterate - Failed to begin transaction");
+            continue;
+        }
+        /* create cursor */
+        rc = dblayer_new_cursor(cldb->be, cldb->db, txnid, &dblcictx->cursor);
+        if (rc != 0) {
+            rc = _cl5Dberror(cldb, rc, "_cl5Iterate - Failed to create cursor");
+            TXN_ABORT(cldb, txnid);
+            continue;
+        }
+        if (startcsn) {
+            char csnstr[CSN_STRSIZE] = "";
+            dbi_val_t startingkey = {0};
+            csn_as_string(startcsn, 0, csnstr);
+            dblayer_value_set(cldb->be, &startingkey, csnstr, CSN_STRSIZE);
+            rc = dblayer_cursor_iterate(&dblcictx->cursor, action_cb, &startingkey, dblcictx);
+        } else {
+            rc = dblayer_cursor_iterate(&dblcictx->cursor, action_cb, NULL, dblcictx);
+        }
+        if (dblcictx->seen.nbmax > 10 && rc == CL5_DB_LOCK_ERROR) {
+            /*
+             * Ran out of locks, need to restart the transaction.
+             * Reduce the the batch count and reset the key to
+             * the starting point.
+             * PR: Not sure weither this code is still working because dbimpl API layer
+             *  may remap bdb error 12 to something else.
+             */
+            dblcictx->seen.nbmax -= 10;
```

Comment:
Were you able to verify that this handling of DB_LOCK_ERROR works ?
If it does that is nice as I was thinking than when the server was getting out of lock, it was almost fatal.
If it works, shouldn't we reduce #seen more significantly ? -10 is just 0.1% step compare to MAX_LOOKUP_PER_TRANSACTION ?

Reply:
Once again 10 was the value in the old code
And IMHO reducing by 10 is fine, because we are not decreasing MAX_LOOKUP_PER_TRANSACTION 
 but the number of operation we were already able to processed (so -1 should be ok).
The only case it should fail a second time is if the other threads are using more txn than it was the case in the first time.
And no, I have not tested, IMHO this feature is broken because bdb_map_error does not handle 12
Should add a new DBI_RC return code to handle cleanly that case.
 and use it instead of the CL5_DB_LOCK_ERROR that we should remove. 
 I am creating a new github issue for that one.

Reply:
Anyway I think that this code is probably obsolete because  the lock monitoring thread should now stop the server before getting out of locks

Reply:
Thanks for the explanations